src/gpu/eigensolver.cpp

```diff
@@ -54,6 +54,27 @@ auto eigh(ContextInternal& ctx, T wl, ConstDeviceView<api::ComplexType<T>, 2> s,
     }
   }
 
+  // Discard rows / columns with sum close to zero
+  for (std::size_t col = 0; col < s.shape(1); ++col) {
```

Comment:
I'd suggest counting the number of non-zero visibilities here as well and return it together with the number of eigenvalues. This way we only iterate through the matrix once. For example by using an `std::pair`.

Reply:
Done.